src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -141,9 +141,12 @@ internal Memory<byte> GetMemory(int sizeHint)
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.minimumSize);
             }
 
-            lock (_sync)
+            // If writing is currently active and enough space, don't need to take the lock to just set WritingActive.
+            // IsWritingActive is needed to prevent the reader releasing the writers memory when it fully consumes currently written.
+            if (!_operationState.IsWritingActive ||
+                _writingMemory.Length == 0 || _writingMemory.Length < sizeHint)
```

Comment:
This is now doing three property reads outside a lock that were presumably previously done in a lock.   Is there a multithreading scenario that can cause a torn state now?

Reply:
No less than before; neither the reader nor the writer are threadsafe individually. They can be used simulatously so the lock is to protect updates that cross from reader->writer or writer->reader.

The `_operationState` is always set under lock as its shared and read by both reader and writer; however it is the writer's flag that's being checked and `_writingMemory` is owned by the writer, so it doesn't need the lock to check the stuff that it sets.

e.g. the writer happily slices and sets `_writingMemory` not under lock in `Advance(int bytesWritten)`